crates/transpiler/src/passes/commutative_optimization.rs

```diff
@@ -0,0 +1,506 @@
+    commutation_checker: &mut CommutationChecker,
+) -> bool {
+    let op1 = inst1.op.view();
+    let op2 = inst2.op.view();
+    let params1 = inst1.params_view();
+    let params2 = inst2.params_view();
+    let qargs1 = dag.get_qargs(inst1.qubits);
+    let qargs2 = dag.get_qargs(inst2.qubits);
+    let cargs1 = dag.get_cargs(inst1.clbits);
+    let cargs2 = dag.get_cargs(inst2.clbits);
+
+    commutation_checker
+        .commute(
+            &op1,
+            params1,
+            qargs1,
+            cargs1,
+            &op2,
+            params2,
+            qargs2,
+            cargs2,
+            u32::MAX,
+            matrix_max_num_qubits,
+            approximation_degree,
+        )
+        .expect("Commutation checker should work")
+}
+
+/// Merge two instructions.
+///
+/// The two instructions have already been canonicalized.
+///
+/// Arguments:
+///
+/// * `dag`: The output [DAGCircuit] that contains all interned qubits.
+/// * `tol`: Specifies tolerance.
+/// * `matrix_max_num_qubits`: The maximum number of qubits to use for more expensive
+///   matrix-based checks.
+///
+/// # Returns:
+///
+/// A triple, consisting of whether the two instructions can be merged, the
+/// merged instruction (`None` if the two instructions cancel out, up to a
+/// global phase), and the global phase update. In other words:
+/// * (true, None, phase_update): the two instructions cancel out, producing
+///   the given global phase.
+/// * (true, Some(instruction), phase_update): the two instructions are merged.
+/// * (false, None, 0.): the two instructions cannot be merged.
+fn try_merge(
+    dag: &DAGCircuit,
+    inst1: &PackedInstruction,
+    inst2: &PackedInstruction,
+    tol: f64,
+    matrix_max_num_qubits: u32,
+) -> PyResult<(bool, Option<PackedInstruction>, f64)> {
+    if inst1.op.num_qubits() != inst2.op.num_qubits() {
+        return Ok((false, None, 0.));
+    }
+
+    let error_cutoff_fn = |_inst: &PackedInstruction| -> f64 { tol };
+
+    let params1 = inst1.params_view();
+    let params2 = inst2.params_view();
+    let qargs1 = dag.get_qargs(inst1.qubits);
+    let qargs2 = dag.get_qargs(inst2.qubits);
+    let cargs1 = dag.get_cargs(inst1.clbits);
+    let cargs2 = dag.get_cargs(inst2.clbits);
+
+    if (qargs1 != qargs2) || (cargs1 != cargs2) {
+        return Ok((false, None, 0.));
+    }
+
+    // Check: both instructions are standard gates which cancel out.
+    if let (OperationRef::StandardGate(gate1), OperationRef::StandardGate(gate2)) =
+        (inst1.op.view(), inst2.op.view())
+    {
+        if let Some((gate1inv, params1inv)) = gate1.inverse(params1) {
+            if (gate1inv == gate2) && compare_params(&params1inv, params2) {
+                return Ok((true, None, 0.));
+            }
+        }
+    }
+    // Check: can merge special standard gates (currently RZ and RX rotations, but we
+    // should include more gates).
+    let merged_gate = match (inst1.op.view(), inst2.op.view()) {
```

Comment:
Could we do this more efficiently for all standard gates if we just check that the operations of `inst1` and `inst2` are the same? I.e. something like
```rust
let merged_gate = if inst1.op.view() == inst2.op.view() { 
  inst1.op.view()
} else { None };
```